### ISSUE TITLE
MySQL Master isn't starting

### ISSUE THREAD

**Issue by muemich** [autopilotpattern/mysql#28]
When I do a docker-compose up I got the following:

```
mysql_1   | 2016-05-09 09:24:40,909 INFO manage.py Initializing database...
mysql_1   | Installing MySQL system tables...2016-05-09 09:24:41 0 [Warning] Using unique option prefix key_buffer instead of key_buffer_size is deprecated and will be removed in a future release. Please use the full name instead.
mysql_1   | 2016-05-09 09:24:41 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
mysql_1   | 2016-05-09 09:24:41 0 [Note] /usr/sbin/mysqld (mysqld 5.6.29-76.2) starting as process 20 ...
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: innodb_empty_free_list_algorithm has been changed to legacy because of small buffer pool size. In order to use backoff, increase buffer pool at least up to 20MB.
mysql_1   | 
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Using atomics to ref count buffer pool pages
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: The InnoDB memory heap is disabled
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Memory barrier is not used
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Compressed tables use zlib 1.2.8
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Using Linux native AIO
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Using CPU crc32 instructions
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Initializing buffer pool, size = 5.0M
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Completed initialization of buffer pool
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: The first specified data file ./ibdata1 did not exist: a new database to be created!
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Setting file ./ibdata1 size to 12 MB
mysql_1   | 2016-05-09 09:24:41 20 [Note] InnoDB: Database physically writes the file full: wait...
mysql_1   | 2016-05-09 09:24:41 7f1fb89da740 InnoDB: Error: Write to file ./ibdata1 failed at offset 0.
mysql_1   | InnoDB: 1048576 bytes should have been written, only 0 were written.
mysql_1   | InnoDB: Operating system error number 28.
mysql_1   | InnoDB: Check that your OS and file system support files of this size.
mysql_1   | InnoDB: Check also that the disk is not full or a disk quota exceeded.
mysql_1   | InnoDB: Error number 28 means 'No space left on device'.
mysql_1   | InnoDB: Some operating system error numbers are described at
mysql_1   | InnoDB: http://dev.mysql.com/doc/refman/5.6/en/operating-system-error-codes.html
mysql_1   | 2016-05-09 09:24:41 20 [ERROR] InnoDB: Error in creating ./ibdata1: probably out of disk space
mysql_1   | 2016-05-09 09:24:41 20 [ERROR] InnoDB: Could not open or create the system tablespace. If you tried to add new data files to the system tablespace, and it failed here, you should now edit innodb_data_file_path in my.cnf back to what it was, and remove the new ibdata files InnoDB created in this failed attempt. InnoDB only wrote those files full of zeros, but did not yet use them in any way. But be careful: do not remove old data files which contain your precious data!
mysql_1   | 2016-05-09 09:24:41 20 [ERROR] Plugin 'InnoDB' init function returned error.
mysql_1   | 2016-05-09 09:24:41 20 [ERROR] Plugin 'InnoDB' registration as a STORAGE ENGINE failed.
mysql_1   | 2016-05-09 09:24:41 20 [ERROR] Unknown/unsupported storage engine: InnoDB
mysql_1   | 2016-05-09 09:24:41 20 [ERROR] Aborting
mysql_1   | 
mysql_1   | 2016-05-09 09:24:41 20 [Note] Binlog end
mysql_1   | 2016-05-09 09:24:41 20 [Note] /usr/sbin/mysqld: Shutdown complete
mysql_1   | 
mysql_1   | FATAL ERROR: Error closing mysqld pipe: Broken pipe
```
I tried to add this to the manage.py:

`self.innodb_data_file_path = get_environ('INNODB_FILEPATH', None)` 
but with or without this the build exits with: 
```
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-FiSa7Z/cffi/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-LlOePZ-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-FiSa7Z/cffi/
The command '/bin/sh -c curl -Ls -o get-pip.py https://bootstrap.pypa.io/get-pip.py &&     python get-pip.py &&     pip install         PyMySQL==0.6.7         python-Consul==0.4.7         manta==2.5.0         mock==2.0.0' returned a non-zero code: 1
 mbp  ~ / projects / autopilot-mysql  master  1  docker images
```

**Comment by tgross** [comment]
You've got this error in the logs above:

> mysql_1   | InnoDB: Error number 28 means 'No space left on device'.

Can you check the disk before we start digging into this further? If that looks good, can you share whether you're using the Compose file and image as-is or whether you've made changes?

**Comment by muemich** [comment]
I saw that as well but
```
df -h
Filesystem      Size   Used  Avail Capacity  iused    ifree %iused  Mounted on
/dev/disk1     465Gi  245Gi  220Gi    53% 64195080 57650454   53%   /
devfs          328Ki  328Ki    0Bi   100%     1136        0  100%   /dev
map -hosts       0Bi    0Bi    0Bi   100%        0        0  100%   /net
map auto_home    0Bi    0Bi    0Bi   100%        0        0  100%   /home
```
I'm using the images as is and the compose file. I added a link to the compose file for mysql to find consul

**Comment by tgross** [comment]
Interesting... what's the deployment environment? This looks like it's on Docker for Mac?

**Comment by muemich** [comment]
```
Filesystem      Size  Used Avail Use% Mounted on
none             60G  2.4G   54G   5% /
tmpfs          1002M     0 1002M   0% /dev
tmpfs          1002M     0 1002M   0% /sys/fs/cgroup
/dev/vda2        60G  2.4G   54G   5% /etc/hosts
shm              64M     0   64M   0% /dev/shm
```
Funny I just deleted the whole sparse image (18G) and rebooted the machine. The max. should be at 64G.
The above mentioned prob seems to be gone, but now I get this:
```
mysql_1   | Version: '5.6.29-76.2-log'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  Percona Server (GPL), Release 76.2, Revision ddf26fe
mysql_1   | 2016-05-09 13:22:22,817 ERROR manage.py mysql_tzinfo_to_sql returned error.
mysql_1   | 2016-05-09 13:22:22,818 INFO manage.py Generated root password: LrUbNKaQFpUrIHnhsAxT
mysql_1   | 2016-05-09 13:22:22 69 [Warning] 'proxies_priv' entry '@ root@e84c8306e2ee' ignored in --skip-name-resolve mode.
mysql_1   | 2016-05-09 13:22:22 69 [Warning] 'proxies_priv' entry '@ root@e84c8306e2ee' ignored in --skip-name-resolve mode.
mysql_1   | 2016-05-09 13:22:22,824 ERROR manage.py (1227, u'Access denied; you need (at least one of) the SUPER, REPLICATION CLIENT privilege(s) for this operation')
```

**Comment by tgross** [comment]
Can you drop that whole log in a gist?

**Comment by muemich** [comment]
https://gist.github.com/muemich/51f5eaa92581bf60bdd034ed34813644

After changing the user/pwd the master came up. But the slaves still aren't working.

**Comment by tgross** [comment]
Just so that I'm clear, the logs you have here are from a totally separate run, right?

You've got [this line](https://gist.github.com/muemich/51f5eaa92581bf60bdd034ed34813644#file-output-L522) in your log for `mysql_2`:

> [32mmysql_2   |[0m 2016-05-09 13:49:27,768 INFO manage.py Initializing database...

That log entry is in the `initialize_db` function, which gets call in the [`pre_start`](https://github.com/autopilotpattern/mysql/blob/88020b92bd2e737b34ca4fd6ace6eb4da95189e4/bin/manage.py#L298-L312) function only when there's no snapshot to fetch from the object store. Because we're initializing the DB in the replicas instead of getting the DB from the snapshot, the replicas don't have the user we need to set up replication.

So the thing we need to figure out is what happened to the snapshot. What jumps out at me in the log are the lines that we're missing. [Somewhere in this section](https://github.com/autopilotpattern/mysql/blob/88020b92bd2e737b34ca4fd6ace6eb4da95189e4/bin/manage.py#L470-L485) it looks like we died.

Here's where we should look next:
- Verify that the snapshot exists on Manta.
- Verify that the LAST_BINLOG_KEY and/or LAST_BACKUP_KEY have been written to Consul.
- If you create a replica with `LOG_LEVEL=DEBUG`, you'll see traces and exit values for most of the functions in this path.

**Comment by muemich** [comment]
Yes they are!

- there is a snapshot on Manta
- the values in consul are called `mysql-last-backup` and `mysql-last-binlog`
- Will try that

But there is a difference in the KEYS you mentioned and the KEYS in consul. So I deleted all backups and tried a new run. And it's working.

All of this is a bit strange. And I can't reproduce all the issues faced

**Comment by tgross** [comment]
> But there is a difference in the KEYS you mentioned and the KEYS in consul.

Sorry, the ALL CAPS keys I mentioned are the consts in the code; they [can be overridden](https://github.com/autopilotpattern/mysql/blob/88020b92bd2e737b34ca4fd6ace6eb4da95189e4/bin/manage.py#L88-L90) in your `_env` file which is why I called them by the const name rather than the specific bit of text.

> And I can't reproduce all the issues faced

Ugh... that's the worst. Not your fault, of course, just the trouble with replicating these kinds of failures. Turns out on second look that the log entires I was looking for in your logs were `log.debug` anyways, so as far as I can tell everything worked as expected from the primary and the failure was at the replicas. Maybe the replicas came up before the snapshot was done being written but once we check Manta/Consul everything is in place? That might be an area to improve.

It looks like there's a couple spots we could hook in more logging too. The balance of debug vs info has been hard to get right; I'm wondering if we should keep the last N messages of a DEBUG level in a ring log with some mechanism to dump them to stdout even if the instances are logging at INFO.

I'm also in the midst of working on an integration test rig so that I can do end-to-end testing on this and other autopilotpattern applications. So hopefully with better testing we'll be able to suss out these issues.
